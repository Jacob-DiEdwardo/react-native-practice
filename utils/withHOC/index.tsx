import React from 'react'

type HOC<IncomingProps, OutgoingProps> = (props: IncomingProps) => OutgoingProps

const withHOC =
  <IncomingProps = {}, OutgoingProps = {}>(
    hoc: HOC<IncomingProps, OutgoingProps>,
    displayName?: string,
  ) => 
  (WrappedComponent: React.FC<IncomingProps & OutgoingProps>) => {
    const Wrapper = (props: IncomingProps) => (
      <WrappedComponent {...props} {...hoc(props)} />
    )
    if (displayName) {
      Wrapper.displayName = displayName
    }
    return Wrapper
  }

export default withHOC
