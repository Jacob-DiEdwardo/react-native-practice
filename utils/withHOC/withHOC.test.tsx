/*
 * @jest-environment jsdom
 */

import { test, expect, afterEach } from '@jest/globals'
import React from 'react'
import { render, cleanup } from '@testing-library/react'
import withHOC from './index'

interface DummyComponentProps {
  test: string
}

afterEach(cleanup)

const DummyComponent: React.FC<DummyComponentProps> = ({ test }) => (
  <div>{test}</div>
)
const withDummyComponent = withHOC(() => ({ test: 'Hello World!' }))
const HOCedDummyComponent = withDummyComponent(DummyComponent)

test('withHOC', () => {
  const tree = render(<HOCedDummyComponent />)
  expect(tree.container).toMatchSnapshot()
})
