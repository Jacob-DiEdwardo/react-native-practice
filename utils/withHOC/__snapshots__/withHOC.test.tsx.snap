// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withHOC 1`] = `
<div>
  <div>
    Hello World!
  </div>
</div>
`;
